ghcide/src/Development/IDE/Core/Debouncer.hs

```diff
@@ -31,28 +31,26 @@ newtype Debouncer k = Debouncer { registerEvent :: Seconds -> k -> IO () -> IO (
 
 -- | Debouncer used in the IDE that delays events as expected.
 newAsyncDebouncer :: (Eq k, Hashable k) => IO (Debouncer k)
-newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> newVar Map.empty
+newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> STM.newIO
 
 -- | Register an event that will fire after the given delay if no other event
 -- for the same key gets registered until then.
---
--- If there is a pending event for the same key, the pending event will be killed.
--- Events are run unmasked so it is up to the user of `registerEvent`
--- to mask if required.
-asyncRegisterEvent :: (Eq k, Hashable k) => Var (HashMap k (Async ())) -> Seconds -> k -> IO () -> IO ()
-asyncRegisterEvent d 0 k fire = do
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Nothing)) k m
-        return (m', cancel)
-    fire
-asyncRegisterEvent d delay k fire = mask_ $ do
-    a <- asyncWithUnmask $ \unmask -> unmask $ do
-        sleep delay
-        fire
-        modifyVar_ d (evaluate . Map.delete k)
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Just a)) k m
-        return (m', cancel)
+asyncRegisterEvent :: (Eq k, Hashable k) => STM.Map k (TVar (Seconds, IO())) -> Seconds -> k -> IO () -> IO ()
+asyncRegisterEvent d delay k fire = join $ atomicallyNamed "debouncer - register" $ do
+    prev <- STM.lookup k d
+    case prev of
+        Just v -> writeTVar v (delay, fire) >> return (pure ())
+        Nothing
+          | delay == 0 -> return fire
```

Comment:
This is a change in semantics, right? Since this would run the action in the calling thread, whereas before it was always run in a background thread.

Reply:
No, the previous code has this special case as well

Reply:
Hmm, I guess it seems a bit suspicious that `asyncRegisterEvent` sometimes runs things synchronously. Seems like it might result in us blocking the main thread when we didn't mean to.

Reply:
It's there to reduce nondeterminism in tests

Reply:
Isn't that what the noop debouncer is for?

Reply:
My last comment was wrong. The delay is set to 0 when reporting on non null diagnostics:

https://github.com/haskell/haskell-language-server/blob/89c44bfaf75326b908548e74ec85eada6c9d3e19/ghcide/src/Development/IDE/Core/Shake.hs#L1168-L1169

And the tests do not run with delay 0 at all, so changing the test suite to use the noop debouncer is likely to result in test failures

Reply:
IDK, it's all a bit indirect. There's a secret behaviour that if you pass 0 delay the event runs not only with no delay but synchronously. I don't really have a better suggestion, though.

And it's not actually for determinism, it's just to... get reports on diagnostics ASAP?